src/kotoba/agent/database.c

```diff
@@ -0,0 +1,330 @@
+/*
+ Author: Harshit Gandhi
+ SPDX-FileCopyrightText: © 2025 Harshit Gandhi <gandhiharshit716@gmail.com>
+
+ SPDX-License-Identifier: GPL-2.0-only
+*/
+
+#define _GNU_SOURCE
+#include <stdio.h>
+
+#include "database.h"
+
+#define LICENSE_REF_TABLE "ONLY license_ref"
+
+PGresult* queryFileIdsForUploadAndLimits(fo_dbManager* dbManager, int uploadId,
+                                         long left, long right, long groupId,
+                                         bool ignoreIrre, bool scanFindings) {
+  char* tablename = getUploadTreeTableName(dbManager, uploadId);
+  gchar* stmt;
+  gchar* sql;
+  PGresult* result;
+
+  char* distinctPfile = "SELECT DISTINCT pfile_fk FROM %s"
+                        " WHERE upload_fk = $1 AND (ufile_mode&x'3C000000'::int) = 0 "
+                        " AND (lft BETWEEN $2 AND $3) AND pfile_fk != 0";
+  char* distinctPfileNoDec = "SELECT DISTINCT pfile_fk FROM ("
+                             "SELECT distinct ON(ut.uploadtree_pk, ut.pfile_fk, scopesort) ut.pfile_fk pfile_fk, ut.uploadtree_pk, decision_type,"
+                             " CASE cd.scope WHEN 1 THEN 1 ELSE 0 END AS scopesort"
+                             " FROM %s AS ut "
+                             " LEFT JOIN clearing_decision cd ON ("
+                             "  (ut.uploadtree_pk = cd.uploadtree_fk AND cd.scope = 0 AND cd.group_fk = $5)"
+                             "  OR (EXISTS ("
+                             "    SELECT 1 FROM report_info ri WHERE ri.upload_fk = ut.upload_fk AND ri.ri_globaldecision = 1"
+                             "  ) AND ut.pfile_fk = cd.pfile_fk AND cd.scope = 1)"
+                             ")"
+                             " WHERE upload_fk=$1 AND (ufile_mode&x'3C000000'::int)=0 AND (lft BETWEEN $2 AND $3) AND ut.pfile_fk != 0"
+                             " ORDER BY ut.uploadtree_pk, scopesort, ut.pfile_fk, clearing_decision_pk DESC"
+                             ") itemView WHERE decision_type!=$4 OR decision_type IS NULL";
+  char* nonVoidPfile = "SELECT pfile_fk FROM allPfileData"
+                       " WHERE pfile_fk NOT IN (SELECT lf.pfile_fk"
+                       " FROM license_file lf"
+                       " JOIN ars_master am ON lf.agent_fk = am.agent_fk"
+                       " JOIN " LICENSE_REF_TABLE " lr ON lf.rf_fk = lr.rf_pk"
+                       " WHERE am.upload_fk = $1"
+                       " AND lr.rf_shortname = ANY(VALUES('No_license_found'), ('Void')))";
```

Comment:
Same here. These functions are already written for you. Just putting the same files again and again in some other folder is not the right way to do it.

https://github.com/fossology/fossology/blob/56f2be3a94b82cd3fca6bf1731ae0ebcee8f1f60/src/monk/agent/database.c#L23-L45


this is true for most of the functions below.  You can refer these functions using CMake from Monk itself. Try and reuse the functions.